addon-resizer/nanny/kubernetes_client.go

```diff
@@ -151,16 +153,54 @@ func (k *kubernetesClient) UpdateDeployment(resources *corev1.ResourceRequiremen
 	// Modify the Deployment object with our ResourceRequirements.
 	for i, container := range dep.Spec.Template.Spec.Containers {
 		if container.Name == k.container {
-			// Update the deployment.
-			dep.Spec.Template.Spec.Containers[i].Resources = *resources
-			_, err = k.clientset.AppsV1().Deployments(k.namespace).Update(dep)
-			return err
+			return k.patchDeployment(getContainerResourcesPatch(i, mergeResources(&container.Resources, resources)))
```

Comment:
Consider what happens if the deployment spec changes between the Get and the Patch request.

Previously, the Update would conflict and force a retry. Now, the patch will be applied to the container at the specified index, even if it is not the expected container.

I'd suggest adding a test to the patch the make sure the name of the container at index %d is still the expected container.

Reply:
Thanks! Makes a lot of sense.

Implemented and tested in all the variants. Looks good.